cmake/DownloadFolly.cmake

```diff
@@ -0,0 +1,39 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+
+## Download facebook's folly library. 
+## SOURCE_DIR is typically the cmake source directory
+## BINARY_DIR is the build directory, typically 'build'
+
+function(download_folly SOURCE_DIR BINARY_DIR)
+
+	
+	ExternalProject_Add(
+		facebook-folly-proj
+		GIT_REPOSITORY "https://github.com/facebook/folly.git"
+		GIT_TAG "v2020.05.18.00"
+		SOURCE_DIR "${BINARY_DIR}/dependencies/facebook-folly-proj-src"
+		CMAKE_ARGS ${PASSTHROUGH_CMAKE_ARGS}
```

Comment:
Interestingly I'm not able to build this git tag, tried it separately too, didn't you run into this? (file name FindFmt vs Findfmt)

```
CMake Error at CMake/folly-deps.cmake:227 (find_package):
  No "Findfmt.cmake" found in CMAKE_MODULE_PATH.
Call Stack (most recent call first):
  CMakeLists.txt:113 (include)


CMake Warning (dev) at CMake/folly-deps.cmake:227 (find_package):
  Findfmt.cmake must either be part of this project itself, in this case
  adjust CMAKE_MODULE_PATH so that it points to the correct location inside
  its source tree.

  Or it must be installed by a package which has already been found via
  find_package().  In this case make sure that package has indeed been found
  and adjust CMAKE_MODULE_PATH to contain the location where that package has
  installed Findfmt.cmake.  This must be a location provided by that package.
  This error in general means that the buildsystem of this project is relying
  on a Find-module without ensuring that it is actually available.

Call Stack (most recent call first):
  CMakeLists.txt:113 (include)
This warning is for project developers.  Use -Wno-dev to suppress it.


```

Reply:
![image](https://user-images.githubusercontent.com/1781585/82658131-7d049800-9bf4-11ea-9c87-804dc9d821d9.png)

I had to trace my steps back but I already had libfmt-dev, which installs the configuration mechanism in (my case ) the cmake root module path. 

Folly/Fizz/Wangle have a lot of dependencies, but if this was the only other one you encountered it would not be difficult to add a similar mechanism for installing libfmt. thanks for pointing this out.

Reply:
Next problem is the "version.h"... 
```
cp: cannot stat '../bin/../../hbase-common/target/generated-sources/native/utils//*': No such file or directory
CMakeFiles/copy_version_h.dir/build.make:57: recipe for target 'CMakeFiles/copy_version_h' failed
make[2]: *** [CMakeFiles/copy_version_h] Error 1
CMakeFiles/Makefile2:67: recipe for target 'CMakeFiles/copy_version_h.dir/all' failed
make[1]: *** [CMakeFiles/copy_version_h.dir/all] Error 2
Makefile:140: recipe for target 'all' failed
make: *** [all] Error 2
```

I don't see hbase-common in the hbase repo generating a native header file. It only generates a Version.java file.. the following code looks sketchy to me. How did get the version.h in your case?

```
# Copy the version.h generated from hbase-common/src/saveVersion.sh script via the mvn build
BIN_DIR=$(dirname "$0")
VERSION_SOURCE_DIR="${BIN_DIR}/../../hbase-common/target/generated-sources/native/utils/"
VERSION_DEST_DIR="${BIN_DIR}/../include/hbase/utils/"
cp $VERSION_SOURCE_DIR/* $VERSION_DEST_DIR/     
```                

(I looked at the git history for that file and it looks like it never generated a header file, I don't know what this is referring to, also the directory paths look wrong).

Reply:
Its probably worth noting, I ran into similar problems (as libfmt) with gmock and gtest. apt install libgtest doesn't install the sharedlibs, I had to manually compile and install again (weird).

Reply:
Yeah version is an issue I'd like to address with cmake getting a known tag. I manually created a version.h for the time being. I started working on a solution for this, but then the size of the PR would grow even more. My hope is to leverage an internal java project that runs a mini cluster and relies on maven to build a jar we can run for integration tests. Right now there is a relative path expecting classpath information to be valid. 

libfmt issues are one I think I need to address. In the case of gtest did you use libgtest-dev? My hope is to control these ( like I started doing with Sodium ). The ultimate issue I would hope to avoid is not being in control of the dependency tree. I think having a build mode for libhabaseclient.so that allows for static linking ( with perhaps GLIB as the only dependency ) will allow better control over not only the dependency tree, but also allow the client to be used across a variety of systems without regard to dependencies on those machines. 

This would also allow python bindings to eventually be a breeze and they can be distributed via pypi with relative ease. 

Finally, RE the copy_version [1] agree it's reliant on a relative path that probably won't exist for most people. A more sustainable solution would be to have a variable specifying an hbase release target, and we generate the version file from the tag. 

I think for the non-draft PR I would probably make the dependency reliance (BUILD_FB_DEPENDENCIES, BUILD_ZOOKEEPER) off by default. That would help to shrink the PR some as the need to download all dependencies and solve every problem is minimized. 

[1] https://github.com/apache/hbase-native-client/blob/master/bin/copy-version.sh

Reply:
> Yeah version is an issue I'd like to address with cmake getting a known tag. I manually created a version.h for the time being. I started working on a solution for this, but then the size of the PR would grow even more. 

> Finally, RE the copy_version [1] agree it's reliant on a relative path that probably won't exist for most people. A more sustainable solution would be to have a variable specifying an hbase release target, and we generate the version file from the tag.

Thanks. To keep things simple, IMHO it is also reasonable to pull the version.h from hbase source in the parent directory (basically the current approach). Just that mvn compilation in hbase project is not generating that header file. Let me know if you think this is a reasonable approach and I can fix it in a separate patch and you can integrate it with your change (that'll keep your changes simple for now).

> My hope is to leverage an internal java project that runs a mini cluster and relies on maven to build a jar we can run for integration tests. Right now there is a relative path expecting class-path information to be valid.

Agreed.

> libfmt issues are one I think I need to address. In the case of gtest did you use libgtest-dev?

Yes, I was using the dev library.  I think this is a known problem, see the second answer on this page.
https://stackoverflow.com/questions/13513905/how-to-set-up-googletest-as-a-shared-library-on-linux

> but also allow the client to be used across a variety of systems without regard to dependencies on those machines.

+1. I think an ideal end state would be a prebuilt toolchain for various commonly used platforms we just download them on the fly during compilation.

Reply:
https://github.com/apache/hbase/pull/1794 is what I was talking about. I couldn't add you to the PR for review but feel free to talk a look. Hope it helps you in this effort.

Reply:
> My hope is to leverage an internal java project that runs a mini cluster and relies on maven to build a jar we can run for integration tests.

This would be great. https://github.com/apache/hbase/tree/master/hbase-testing-util should have "enough" of what's needed to get going. Shout when you're ready for this and we can help. The hardest part will be explicitly listing all of the test scope dependencies, as maven won't naturally bring them in transitively (super fun -- test-scope deps have no transitive deps).

> If we end up dockerizing this on ubuntu, we may run into this problem in which case we may have to install fmt dependency like other dependencies, just FYI.

I think Docker is going to be a "requirement" for the rest of the HBase team to try to be "helpful". I'm not sure if there are any frameworks out there which would help us do the same setup+build+test across multiple OS's docker containers. That said, having just "one" platform which works is more than we have now :)

Reply:
Totally agree. I'm going to submit a separate PR where it should have just the cmake changes without changing dependencies. Since I already have these changes that should prep a separate PR for that change set. For now I'm working through compilation issues on the older folly.